service/discovery/aws/compute.go

```diff
@@ -0,0 +1,243 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *           $$\                           $$\ $$\   $$\
+ *           $$ |                          $$ |\__|  $$ |
+ *  $$$$$$$\ $$ | $$$$$$\  $$\   $$\  $$$$$$$ |$$\ $$$$$$\    $$$$$$\   $$$$$$\
+ * $$  _____|$$ |$$  __$$\ $$ |  $$ |$$  __$$ |$$ |\_$$  _|  $$  __$$\ $$  __$$\
+ * $$ /      $$ |$$ /  $$ |$$ |  $$ |$$ /  $$ |$$ |  $$ |    $$ /  $$ |$$ | \__|
+ * $$ |      $$ |$$ |  $$ |$$ |  $$ |$$ |  $$ |$$ |  $$ |$$\ $$ |  $$ |$$ |
+ * \$$$$$$\  $$ |\$$$$$   |\$$$$$   |\$$$$$$  |$$ |  \$$$   |\$$$$$   |$$ |
+ *  \_______|\__| \______/  \______/  \_______|\__|   \____/  \______/ \__|
+ *
+ * This file is part of Clouditor Community Edition.
+ */
+
+package aws
+
+import (
+	"clouditor.io/clouditor/api/discovery"
+	"clouditor.io/clouditor/voc"
+	"context"
+	"errors"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/aws/aws-sdk-go-v2/service/lambda"
+	types2 "github.com/aws/aws-sdk-go-v2/service/lambda/types"
+	"github.com/aws/smithy-go"
+)
+
+// computeDiscovery handles the AWS API requests regarding the computing services (EC2 and Lambda)
+type computeDiscovery struct {
+	virtualMachineAPI EC2API
+	functionAPI       LambdaAPI
+	isDiscovering     bool
+	awsConfig         *Client
+}
+
+// EC2API describes the EC2 api interface (for mock testing)
+type EC2API interface {
+	DescribeInstances(ctx context.Context,
+		params *ec2.DescribeInstancesInput,
+		optFns ...func(options *ec2.Options)) (*ec2.DescribeInstancesOutput, error)
+}
+
+// LambdaAPI describes the lambda api interface (for mock testing)
+type LambdaAPI interface {
+	ListFunctions(ctx context.Context,
+		params *lambda.ListFunctionsInput, optFns ...func(*lambda.Options)) (*lambda.ListFunctionsOutput, error)
+}
+
+// newFromConfigEC2 holds ec2.NewFromConfig(...) allowing a test function to mock it
+var newFromConfigEC2 = ec2.NewFromConfig
+
+// newFromConfigLambda holds lambda.NewFromConfig(...) allowing a test function tp mock it
+var newFromConfigLambda = lambda.NewFromConfig
+
+// NewComputeDiscovery constructs a new awsS3Discovery initializing the s3-virtualMachineAPI and isDiscovering with true
+func NewComputeDiscovery(client *Client) discovery.Discoverer {
+	return &computeDiscovery{
+		virtualMachineAPI: newFromConfigEC2(client.Cfg),
+		functionAPI:       newFromConfigLambda(client.Cfg),
+		isDiscovering:     true,
+		awsConfig:         client,
+	}
+}
+
+// Name is the method implementation defined in the discovery.Discoverer interface
+func (d computeDiscovery) Name() string {
+	return "AWS Compute"
+}
+
+// List is the method implementation defined in the discovery.Discoverer interface
+func (d computeDiscovery) List() (resources []voc.IsCloudResource, err error) {
+	listOfVMs, err := d.discoverVirtualMachines()
+	if err != nil {
+		return
```

Comment:
```suggestion
		return nil, fmt.Errorf("could not discover virtual machines: %w", err)
```

I would wrap the error here, so one knows that it failed while listing virtual machines